### PR TITLE
Psalm integration

### LINKED ISSUE #7: Psalm integration
### Feature Request

|    Q        |   A
|------------ | ------
| QA    | yes

#### Summary

As decided during the [Technical-Steering-Committee Meeting on August 3rd, 2020](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-static-analysis-tooling), Laminas wants to implement [vimeo/psalm](https://github.com/vimeo/psalm) in all packages.

Implementing psalm is quite easy.

#### Required
- [ ] Create a `psalm.xml` in the project root
- [ ] Copy and paste the contents from this [psalm.xml.dist](https://github.com/mezzio/repo-template/blob/6124090610ef1c6d6b0c80a063cafacc3ec68f38/psalm.xml.dist)
- [ ] Run `$ composer require --dev vimeo/psalm`
- [ ] Run `$ vendor/bin/psalm --set-baseline=psalm-baseline.xml`
- [ ] Add a composer script `static-analysis` with the command `psalm --shepherd --stats`
- [ ] Add a new line to `script:` in `.travis.yml`: `- if [[ $TEST_COVERAGE == 'true' ]]; then composer static-analysis ; fi`
- [ ] Remove phpstan from the project (`phpstan.neon.dist`, `.travis.yml` entry, `composer.json` `require-dev` and `scripts`)

##### Optional
- [ ] Fix as many psalm errors as possible.

### PR DESCRIPTION
**Pull request by ghostwriter**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Resolves #7

### CODE REVIEW

File: .github/workflows/psalm.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+name: "Static Analysis by Psalm"
```
Comment (boesing) [inline .github/workflows/psalm.yml]: I'd suggest removing this workflow in favor of the laminas continous integration workflow:
https://gist.github.com/weierophinney/9decd19f76b7d9745c6559074053fa65

File: src/LazyParameterPostProcessor.php [hunk 2]
```diff
@@ -11,22 +11,23 @@ final class LazyParameterPostProcessor
     private $parameterProvider;
 
     /**
-     * @psalm-param callable():array<string,mixed> $parameterProvider
+     * @psalm-param callable():array<array-key, mixed> $parameterProvider
```
Comment (boesing) [inline src/LazyParameterPostProcessor.php]: This is less accurate than `array<string,mixed>`.
Please revert this change as parameters **MUST** have string keys.

File: src/LazyParameterPostProcessor.php [hunk 3]
```diff
@@ -11,22 +11,23 @@ final class LazyParameterPostProcessor
...
      */
     public function __construct(callable $parameterProvider)
     {
         $this->parameterProvider = $parameterProvider;
     }
 
     /**
-     * @param array<string,mixed> $config
+     * @param array<array-key, mixed> $config
```
Comment (boesing) [inline src/LazyParameterPostProcessor.php]: Please revert this change as parameters MUST have string keys.

File: src/LazyParameterPostProcessor.php [hunk 4]
```diff
@@ -11,22 +11,23 @@ final class LazyParameterPostProcessor
...
      *
-     * @return array<string,mixed>
+     * @return array<array-key, mixed>
```
Comment (boesing) [inline src/LazyParameterPostProcessor.php]: Please revert this change as the configuration returned by this configuration **MUST** have string keys.

File: src/ParameterNotFoundException.php [hunk 5]
```diff
@@ -15,23 +15,19 @@
...
-        ), $e->getCode(), $e);
-        $toReturn->key = $e->getKey();
-        return $toReturn;
-    }
+        /** @var string $key */
+        $key = $exception->getKey();
+        /** @var int $code */
+        $code = $exception->getCode();
 
-    public function getKey() : string
```
Comment (boesing) [inline src/ParameterNotFoundException.php:33]: By removing the `getKey` method, you are introducing BC break. Please revert this removal in order to preserve this method.

File: src/ParameterPostProcessor.php [hunk 6]
```diff
@@ -28,14 +28,19 @@ public function __construct(array $parameters)
         $this->parameters = $parameters;
     }
 
+    /**
+     * @return (array|mixed)[]
```
Comment (boesing) [inline src/ParameterPostProcessor.php]: Applying suggested changes from psalm is not helpful.
Please remove this change as it adds 0 value.

File: src/ParameterPostProcessor.php [hunk 7]
```diff
@@ -28,14 +28,19 @@ public function __construct(array $parameters)
...
+     *
+     * @psalm-return array{parameters: array}
```
Comment (boesing) [inline src/ParameterPostProcessor.php:34]: This is not correct. The return value of this method returns more than just the parameters.
Please use the suggested annotation instead.
```suggestion
     * @psalm-return array<string,mixed>
```

File: src/ParameterPostProcessor.php [hunk 8]
```diff
@@ -28,14 +28,19 @@ public function __construct(array $parameters)
...
+     */
     public function __invoke(array $config) : array
     {
 
         try {
             $parameters = $this->getResolvedParameters();
 
-            array_walk_recursive($config, function (&$value) use ($parameters) {
-                $value = $parameters->unescapeValue($parameters->resolveValue($value));
+            array_walk_recursive($config, function (string &$value) use ($parameters) {
```
Comment (boesing) [inline src/ParameterPostProcessor.php]: Please do not cast any value in the config to a string.
Parameters can be `float`, `boolean`, `integer`, `array`, `null` (so just *any* type).
Casting to `string` breaks almost any implementation.

File: src/ParameterPostProcessor.php [hunk 9]
```diff
@@ -49,13 +54,15 @@ public function __invoke(array $config) : array
     private function resolveNestedParameters(array $values, string $prefix = '') : array
     {
         $convertedValues = [];
+        /** @var string|array $value */
```
Comment (boesing) [inline src/ParameterPostProcessor.php]: The value is mixed here and that is absolutely okay as parameter values can have almost any type.
If psalm complains about it, feel free to suppress the error by using `@psalm-suppress`

File: src/ParameterPostProcessor.php [hunk 10]
```diff
@@ -49,13 +54,15 @@ public function __invoke(array $config) : array
...
         foreach ($values as $key => $value) {
             // Do not provide numeric keys as single parameter
             if (is_numeric($key)) {
                 continue;
             }
 
             $convertedValues[$prefix . $key] = $value;
+
```
Comment (boesing) [inline src/ParameterPostProcessor.php:67]: Please try to avoid unrelated changes

File: test/ParameterPostProcessorTest.php [hunk 11]
```diff
@@ -19,7 +19,12 @@ class ParameterPostProcessorTest extends TestCase
 {
     use ArraySubsetAsserts;
 
-    public function parameterProvider()
+    /**
+     * @return (string|string[])[][][]
```
Comment (boesing) [inline test/ParameterPostProcessorTest.php]: You should avoid applying psalm suggestions as these do not provide any helpful value for consumers/maintainers.
The structure of the returned array is 

```
@psalm-return array<string,array{0:array<string,mixed>,1:array{config:array{param:string,used_parameter:string}},2:array{param:mixed,used_parameter:string}}>
```

As you can see, this is ma little bit less detailed **but** if we have to add new unit tests here, we do not have to refactor the whole return type. So if we have an integer parameter we want to test, the return annotation is still valid!

File: test/ParameterPostProcessorTest.php [hunk 12]
```diff
@@ -63,23 +68,25 @@ public function parameterProvider()
 
     /**
      * @dataProvider parameterProvider
+     *
+     * @return void
```
Comment (boesing) [inline test/ParameterPostProcessorTest.php]: There is no need to annotate this twice. native return type is absolutely okay as you've done in all the other tests in the same file.

**Review by boesing — CHANGES_REQUESTED** [review]

File: src/ParameterPostProcessor.php (hunk 8 above)
Comment (ghostwriter) [reply src/ParameterPostProcessor.php]: Shouldn't the tests have failed to indicate this?

Comment (boesing) [reply src/ParameterPostProcessor.php]: yes, but obviously these kind of tests are missing. 
Gonna create an issue for that.

Comment (boesing) [reply src/ParameterPostProcessor.php]: Created #11, thanks for pointing this out.

**Comment by ghostwriter** [comment]
@boesing ping for final review.

File: test/ParameterPostProcessorTest.php [hunk 13]
```diff
@@ -19,7 +19,10 @@ class ParameterPostProcessorTest extends TestCase
 {
     use ArraySubsetAsserts;
 
-    public function parameterProvider()
+    /**
+     * @psalm-return array{root-scoped-parameter: array{0: array{foo: string}, 1: array{config: array{param: string, used_parameter: string}}, 2: array{config: array{param: string, used_parameter: string}}}, multi-level-parameter: array{0: array{foo: array{bar: string}}, 1: array{config: array{param: string, used_parameter: string}}, 2: array{config: array{param: string, used_parameter: string}}}}
```
Comment (nathane) [inline test/ParameterPostProcessorTest.php:23]: CS failure due to `@psalm-return` DocBlock.

**Comment by boesing** [comment]
@ghostwriter Please focus on the PR title. 
I've written a quite long comment in the `laminas-cache-storage-adapter-test` package and now closing this.
Having PRs like these are hard to review. Feel free to seperate your PRs into smaller PRs.

1. Psalm integration (please do not remove already typed return values in any way and try to avoid psalms "alter" logic as it adds too much noise)
2. Codestandard 2.0 integration
3. Composer normalization

**Comment by ghostwriter** [comment]
@boesing Too lazy to click on 18 individual commits, that elegantly encapsulate each action taken? 

Even on a mobile device, GitHub displays all changes that occurred in a single commit separately.

Saying my commits are “impossible to review” is deliberate gross exaggeration.

Assuming you all (laminas org) followed the same guidelines, I was just trying to add as much value to the PR, to help expedite things by consolidating multiple “Features” into a single PR, following @weierophinney’s footsteps (who did the exact same thing I did in this PR, to not one but two separate PRs that I submitted so far.

That said, 
- **I’ll correct this going forward.**
- I’ll `git cherry-pick` the commits of this PR and make multiple PRs during my next available “free” time.

Have a great weekend!

P.S. Regarding Psalm I only changed what was originally there because it was complaining about it being incorrect. Now it’s not.

P.P.S Regarding the other PR you mentioned: Why did you merge a PR that you expertly said was “impossible to review”, leading the reader to believe that you did not review it? What if it had malicious code?

**Pull request closed without merging**